File: python/src/day15/main.py
from src.common.file_utils import get_path, read_integers


def part_one(nums: list) -> int:
    return play_game(nums, 2020)


def part_two(nums: list) -> int:
    return play_game(nums, 30000000)


def play_game(nums: list, turns: int) -> int:
    history = {}
    last_number, spoken = None, None

    for i in range(turns):
        if i < len(nums):
            spoken = nums[i]
        else:
            if last_number not in history:
                spoken = 0
            else:
                spoken = i - 1 - history[last_number]

        history[last_number] = i - 1
        last_number = spoken

    return last_number


if __name__ == '__main__':

    print("---Part One---")
    part_one_res = part_one([7, 14, 0, 17, 11, 1, 2])
    print(part_one_res)

    print("---Part Two---")
    part_two_res = part_two([7, 14, 0, 17, 11, 1, 2])
    print(part_two_res)

File: go/day15/main.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println("--- Part One ---")
	fmt.Println(Part1([]int{7, 14, 0, 17, 11, 1, 2}))

	fmt.Println("--- Part Two ---")
	fmt.Println(Part2([]int{7, 14, 0, 17, 11, 1, 2}))
}

func Part1(nums []int) int {
	return playGame(nums, 2020)
}

func Part2(nums []int) int {
	return playGame(nums, 30000000)
}

func playGame(nums []int, turns int) int {
	a := make([]int, turns)
	seen := make(map[int]int)

	for i, n := range nums {
		a[i] = n
		seen[n] = i
	}

	for n := len(nums); n < turns-1; n++ {
		if m, ok := seen[a[n]]; ok {
			a[n+1] = n - m
		}
		seen[a[n]] = n
	}

	return a[turns-1]
}
